driver-core/src/test/unit/com/mongodb/AuthConnectionStringTest.java

```diff
@@ -93,31 +92,65 @@ private void testValidUris() {
         }
 
         MongoCredential credential = connectionString.getCredential();
-        assertString("auth.db", credential.getSource());
-        assertString("auth.username", credential.getUserName());
+        if (credential != null) {
+            assertString("credential.source", credential.getSource());
+            assertString("credential.username", credential.getUserName());
+            assertMechanismProperties(credential);
+
+            // Passwords for certain auth mechanisms are ignored.
+            String password = credential.getPassword() != null ? new String(credential.getPassword()) : null;
+            if (password != null) {
+                assertString("credential.password", password);
+            }
+
+            assertMechanism("credential.mechanism", credential.getMechanism());
+        }
+    }
 
-        // Passwords for certain auth mechanisms are ignored.
-        String password = credential.getPassword() != null ? new String(credential.getPassword()) : null;
-        if (password != null) {
-            assertString("auth.password", password);
+    private void assertString(final String key, final String actual) {
+        BsonValue expected = getExpectedValue(key);
+
+        if (expected.isNull()) {
+            assertTrue(String.format("%s should be null", key), actual == null);
+        } else if (expected.isString()) {
+            String expectedString = expected.asString().getValue();
+            assertTrue(String.format("%s should be %s but was %s", key, actual, expectedString), actual.equals(expectedString));
+        } else {
+            assertTrue(String.format("%s should be %s but was %s", key, actual, expected), false);
         }
-        if (definition.get("options").isDocument()) {
-            for (Map.Entry<String, BsonValue> option : definition.getDocument("options").entrySet()) {
-                if (option.getKey().equals("authmechanism")) {
-                    String expected = option.getValue().asString().getValue();
-                    if (expected.equals("MONGODB-CR")) {
-                        assertNotNull(connectionString.getCredential());
-                        assertNull(connectionString.getCredential().getAuthenticationMechanism());
+    }
+
+    private void assertMechanism(final String key, final String actual) {
+        BsonValue expected = getExpectedValue(key);
+
+        if (expected.isString() && expected.asString().getValue().equals("MONGODB-CR")) {
```

Comment:
Why is this special case needed?

Reply:
We have the following code in `ConnectionString#createCredentials`:
```
            if (key.equals("authmechanism")) {
                if (value.equals("MONGODB-CR")) {
                    if (userName == null) {
                        throw new IllegalArgumentException("username can not be null");
                    }
                    LOGGER.warn("Deprecated MONGDOB-CR authentication mechanism used in connection string");
                } else {
                    mechanism = AuthenticationMechanism.fromMechanismName(value);
                }
            }
```
The test expects the value for the mechanism will be `MONGODB-CR`, but the Java driver will assign `null`.

Reply:
Right, it's because we removed MONGODB-CR from the AuthenticationMechanism enum for the 4.0 release.

Reply:
This deserves a comment explaining why we need this condition.

Reply:
Sorry, I resolved this comment prematurely.

Reply:
I have added a comment for this condition.